### PR TITLE
Add Single.defer()

### PR DESCRIPTION
**Pull request by artem-zinnatullin**

### CODE REVIEW

File: src/main/java/rx/Single.java [hunk 1]
```diff
@@ -1898,4 +1898,48 @@ public void onNext(T t) {
...
+
+                try {
+                    single = singleFactory.call();
+                } catch (Throwable t) {
+                    Exceptions.throwIfFatal(t);
+                    singleSubscriber.onError(t);
+                    return;
+                }
+
+                single.subscribe(singleSubscriber);
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Please notice that I'm not wrapping subscriber via `Subscribers.wrap()` because `SingleSubscriber` differs from `Subscriber`, that's why I used `subscribe` instead of `unsafeSubscribe()` here.

If this should be changed, please comment.

**Comment by akarnokd** [comment]
Please add unit test of normal, callback throws and callback returns null cases.

**Comment by artem-zinnatullin** [comment]
@akarnokd added `deferShouldPassNullPointerExceptionToTheSubscriberIfSingleFactoryIsNull` and `deferShouldPassNullPointerExceptionToTheSubscriberIfSingleFactoryReturnsNull`.

Btw, I've used `Callable` instead of `Func0` to save users from dealing with checked exception, hope it's okay :)

**Comment by akarnokd** [comment]
Func0 extends Callable so its okay.

.:+1:

**Comment by abersnaze** [comment]
:+1:

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/rx/Single.java]: `SingleSubscriber` doesn't have `onStart` like `Subscriber` so this is not a problem.

**Comment by akarnokd** [comment]
I'm still :+1: but please rebase.

**Comment by artem-zinnatullin** [comment]
Rebased!

**Comment by akarnokd** [comment]
Thanks! Merged.

**Comment by passsy** [comment]
thanks
